### PR TITLE
Document List.filter_map as tailrec; add rev version

### PR DESCRIPTION
**Pull request by thomassa**
Also the same for ListLabels.

### CODE REVIEW

File: stdlib/list.mli [hunk 1]
```diff
@@ -191,10 +191,19 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
...
-    the [Some] elements.
+    the [Some] elements. It preserves the order of the elements (by
+    means of an internal reversal). It is tail-recursive.
     @since 4.08.0
  *)
 
+val rev_filter_map : ('a -> 'b option) -> 'a list -> 'b list
+(** [rev_filter_map f l] gives the same result as
+    {!rev}[ (]{!filter_map}[ f l)], but is more efficient:
+    [filter_map f l] is implemented as [rev rev_filtermap f l].
```
Comment (yallop) [inline stdlib/list.mli]:
```suggestion
    [filter_map f l] is implemented as [rev (rev_filter_map f l)].
```

File: stdlib/listLabels.mli [hunk 2]
```diff
@@ -191,10 +191,19 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
...
-    the [Some] elements.
+    the [Some] elements. It preserves the order of the elements (by
+    means of an internal reversal). It is tail-recursive.
     @since 4.08.0
  *)
 
+val rev_filter_map : f:('a -> 'b option) -> 'a list -> 'b list
+(** [rev_filter_map ~f l] gives the same result as
+    {!rev}[ (]{!filter_map}[ ~f l)], but is more efficient:
+    [filter_map ~f l] is implemented as [rev rev_filtermap ~f l].
```
Comment (yallop) [inline stdlib/listLabels.mli]:
```suggestion
    [filter_map ~f l] is implemented as [rev (rev_filter_map ~f l)].
```

**Review by yallop — CHANGES_REQUESTED** [review]

**Comment by thomassa** [comment]
Thanks @yallop; I'd spotted the missing parens and underscore in the comment myself and was just fixing it when you pointed it out.
I've run `make html_doc` and the output looks right to me in a browser now.

**Comment by thomassa** [comment]
The version in `@since 4.13.0` in the comment will depend on when it's merged to trunk, but I think this is right at present, yes?

**Comment by thomassa** [comment]
```
ocaml/testsuite$ make one DIR=tests/lib-list
Running tests from 'tests/lib-list' ...
 ... testing 'test.ml' with 1 (native) => passed
 ... testing 'test.ml' with 2 (bytecode) => passed
make[1]: Entering directory 'ocaml/testsuite'
make[1]: Leaving directory 'ocaml/testsuite'
make[1]: Entering directory 'ocaml/testsuite'

Summary:
    2 tests passed
    0 tests skipped
    0 tests failed
    0 tests not started (parent test skipped or failed)
    0 unexpected errors
    2 tests considered
make[1]: Leaving directory 'ocaml/testsuite'
```

**Comment by dra27** [comment]
> The version in `@since 4.13.0` in the comment will depend on when it's merged to trunk, but I think this is right at present, yes?

4.13 is frozen, so it should be 4.14, thanks.

**Comment by thomassa** [comment]
@dra27 wrote:
> 4.13 is frozen, so it should be 4.14, thanks.

Updated to 4.14.0 now: thank you.
Meanwhile the appveyor build succeeded on the version that said 4.13: https://ci.appveyor.com/project/avsm/ocaml/builds/40661504

**Comment by Octachron** [comment]
Adding only `rev_filter_map` seems inconsistent (what about `rev_filter` for instance ?). I am not sure how useful the function is in general.

File: stdlib/list.mli (hunk 1 above)
Comment (Octachron) [reply stdlib/list.mli]: I think it is even better to not describe implementation details that could change between version.

File: stdlib/list.mli [hunk 3]
```diff
@@ -191,10 +191,19 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
 val filter_map : ('a -> 'b option) -> 'a list -> 'b list
 (** [filter_map f l] applies [f] to every element of [l], filters
     out the [None] elements and returns the list of the arguments of
-    the [Some] elements.
+    the [Some] elements. It preserves the order of the elements (by
+    means of an internal reversal). It is tail-recursive.
```
Comment (Octachron) [inline stdlib/list.mli:195]: Documenting that the order of elements is preserved and that the function is tail-recursive does seem useful. I am not fond of the parenthetical description of implementation details however.

**Comment by v-gb** [comment]
There are 3 changes here:
- documenting that `List.filter_map` is tail recursive. In trunk, every List function is assumed tail-recursive unless the doc says otherwise, so I don't think we want this change at this point
- documenting that List.filter_map preserves order. Makes sense.
- adding `List.rev_filter_map`. That seems undesirable (because almost no function has a `rev_` equivalent. Why is `List.rev (List.filter_map ...)` not good enough?`List.rev_map` is the exception, not the common case), so you'd need to argue for why this is useful.

So I suggest reducing this pull request to just documenting the order of the `List.filter_map`.

**Comment by dra27** [comment]
~~That's not all strictly true: there are several other instances where a function is noted as being tail-recursive. I can see the case here - `map` is explicitly documented as being not tail recursive, so it is noteworthy that `filter_map` is tail recursive - in particular, `concat_map` already notes this. `List.rev (List.filter_map ...)` features two list reversals, so there's an efficiency argument for having `filter_rev_map`, just as for `rev_map` and `rev_map2` (and also for `rev_append`; `map` is not the exception).~~

~~I'm not against the function, although there is the consideration for where it stops (should we also have `fold_left_rev_map`, for example, and the `partition` functions likewise finish with an accumulator reversal?). @Octachron - what's your feeling on the function?~~ _read the wrong branch..._

**Comment by nojb** [comment]
> `map` is explicitly documented as being not tail recursive,

`map` is tail-recursive since #11362. 

> `List.rev (List.filter_map ...)` features two list reversals,

`filter_map` does not reverse the list since #11402.

> so there's an efficiency argument for having `filter_rev_map`, just as for `rev_map` and `rev_map2` (and also for `rev_append`; `map` is not the exception).

In my mind, the `rev_` functions made sense when the "usual" functions were not tail-recursive, but since TRMC came into the picture, I don't see much use for them (except as micro-optimizations). So, personally, am not sure it is a good idea to add more `rev_` functions to the standard library.

**Comment by xavierleroy** [comment]
> In my mind, the rev_ functions made sense when the "usual" functions were not tail-recursive, but since TRMC came into the picture, I don't see much use for them

Agreed.  I don't think it's worth complicating the List API with more `_rev` functions.

**Comment by dra27** [comment]
I'm so sorry for the noise - I was accidentally reading the 5.0 branch, not trunk 🤦

**Comment by thomassa** [comment]
Sorry: I'd almost forgotten about this pull request: shortly after I opened it, the organisation where I was working at the time switched away from OCaml.

Now that we have the TRMC feature (and a `List` module that uses it where applicable) I think this pull request is largely obsolete; I seem to remember I wrote it after helping fix a stack-overflow bug where a colleague had used something like `xs |> List.filter f |> List.map g |> List.rev` so I had needed to check which of this module's functions were tail-recursive vs. non-tail recursive: `rev_filter_map` was the overall operation to be done, we needed to avoid non-tail recursion, and I wanted to avoid the pointless double-reversal done by `xs |> List.filter_map f |> List.rev` (given that `filter_map` at the time was the old tail-recursive pre-TRMC implementation which applied a `rev` at the end).

Nowadays, the `rev_filter_map` function is less broadly useful, since its main benefit is (or rather was) its speed relative to the old tail-recursive pre-TRMC `filter_map`. (It is still faster than `List.rev (List.filter_map f l)` but by a smaller margin now, and that use-case is probably not common enough to justify "complicating the List API with more `_rev` functions," as @xavierleroy put it: I agree with Xavier and @nojb.)

In the absence of `rev_filter_map` there is no real need for `filter_map` to specify that it preserves the order of the elements.

As for documenting `filter_map` to be tail-recursive, the usefulness was always questionable, since the module-level preamble says that all the functions are tailrec except where noted otherwise. The value is (or was) that someone trying to check a particular function might read just that function's documentation (not seeing the module-level description) and thus not find the wanted information (and then perhaps resort to reading the implementation). On the other hand, the current version of the module (using TRMC) has fewer non-tail recursive functions, especially ones that have natural counterparts that are tail-recursive and reversing (e.g. `map` is now tailrec due to TRMC). Thus there is less need for people to refer to the documentation for questions like "Hold on, have I remembered correctly which is/are tail-recursive out of `map`, `filter`, and `filter_map`, and which of those have tail-recursive `rev` counterparts?"

Therefore I am closing this pull-request.

**Pull request closed without merging**